core/src/main/scala/org/apache/spark/util/LargeByteBufferInputStream.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.io.InputStream
+
+import org.apache.spark.network.buffer.LargeByteBuffer
+import org.apache.spark.storage.BlockManager
+
+/**
+ * Reads data from a LargeByteBuffer, and optionally cleans it up using buffer.dispose()
+ * at the end of the stream (e.g. to close a memory-mapped file).
+ */
+private[spark]
+class LargeByteBufferInputStream(private var buffer: LargeByteBuffer, dispose: Boolean = false)
+  extends InputStream {
+
+  override def read(): Int = {
+    if (buffer == null || buffer.remaining() == 0) {
+      cleanUp()
+      -1
+    } else {
+      val r = buffer.get() & 0xFF
+      if (buffer.remaining() == 0) {
+        cleanUp()
```

Comment:
If anyway is watching on the sidelines -- marcelo and I chatted about this a while and realized there is an issue with the existing use of `ByteBufferInputStream` (where this code was copied from) that prevents it from getting properly disposed in all cases.  I've opened https://issues.apache.org/jira/browse/SPARK-6839